7.9.x Update to jinja2 2.11.1

kinow opened the pull request:
These changes close #2572

Sibling PR of #3493

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered by existing tests.
- [x] Appropriate change log entry included.
- [x] No documentation update required.

kinow commented on tests/validate/43-jinja2-template-error-main.t:
```diff
@@ -33,7 +33,7 @@ __SUITERC__
 contains_ok "${TEST_NAME_BASE}.stderr" <<'__ERROR__'
 Jinja2Error:
-    'You can only sort by either "key" or "value"'
+    raise FilterArgumentError('You can only sort by either "key" or "value"')
```
The other changes are the result of copy-and-paste of Jinja2 2.11.1 ZIP contents, minus their asyncio files as that breaks the build due to Py3 code.

After copying the files, this was the only test that failed. Debugging it, it looks like the traceback contents changed in 2.11.1, and instead of having just the text on a new line, it contains the `raise FilterArgumentError` part too.

Wondering if that's safe to update... I don't think users rely on this?

hjoliver replied:
Seems fine to me.

hjoliver replied:
(What were the asyncio files for though?   Is it surprising that deleting those didn't entirely break the new Jinja2?)

oliver-sanders replied:
Curious, 2.11.1 claims to have 2.7 support. What aspect of the build was breaking? Was is our tests (e.g. pycodestyle) or something functional?

kinow commented:
Compared the output of the new Jinja2 library side-by-side with 7.8.x and master PR branches, and the traceback changed a bit.

![image](https://user-images.githubusercontent.com/304786/74615177-06827500-5183-11ea-8c91-180d9e62c3ff.png)

For the 7.8.x one test failed as instead of starting a new line with `    'You can only...'`, it starts now with `    raise FilterArgumentError('You can only...')`.

![image](https://user-images.githubusercontent.com/304786/74615189-2dd94200-5183-11ea-8eee-66d5a38e0d48.png)

hjoliver commented:
@kinow - well done getting this to work with 2.7

@oliver-sanders -  this fixes a known bug, but are you particularly concerned about Jinja2 back-compat for Met Office Cylc-7 users? (Given that we bundle Jinja2 with Cylc 7, I'm not sure [this statement](https://github.com/cylc/cylc-flow/pull/3493#issuecomment-586795139) is as true for Cylc 7 as it is for Cylc 8...)

oliver-sanders commented:
> this fixes a known bug, but are you particularly concerned about Jinja2 back-compat for Met Office Cylc-7 users?

Jinja2 bugs in workflows are nasty to debug, worse still as workflows often have highly-branched Jinja2 code bugs go unnoticed until a later date by which time the relation to the change of Cylc version is no longer obvious.

So Jinja2 changes need to be well managed and documented to give users a fighting chance. Sneaking a Jinja2 upgrade into a maintenance release would be unhelpful. It would break the trust of our release schedule if maintenance releases have significant interface changes.

From a change management perspective `8.0.0` is a more natural home for this change. Otherwise we could bump `7.8.5` to `7.9`.

oliver-sanders commented:
Second thoughts, from the MO perspective we might still require Python2.6 compatibility for internal release which this would break. Will check.

hjoliver commented:
> From a change management perspective 8.0.0 is a more natural home for this change. But as 8.0.0 is so far off we could bump 7.8.5 to 7.9 and make a (small) minor release out of this.

Can do. Is that OK with you @kinow (so your work here won't be wasted :+1: )   (For background, there are many well-known and well-used Cylc-7 suite configs that are essentially massive Jinja2 "programs" :grimacing: )

hjoliver commented:
> we might still require Python2.6 compatibility for internal release which this would break.

Ouch.

hjoliver commented:
(OK awaiting comment from @dpmatthews ... from offline chat with @oliver-sanders we could put this PR to 7.9.0, and maintain 7.8.x as well for the few remaining bug fixes...)

kinow commented:
Happy if this is postponed to 7.9 or targeted only for 8.0. Whichever works best.

dpmatthews commented:
I don't think we can remove Python 2.6 compatibility in a 7.8.x release so this would need to become 7.9. We still have Cylc in use on RHEL 6 systems so would have to stick to 7.8. Unless there is demand from others, I suggest we make this an 8.0 change.

TomekTrzeciak commented:
> I don't think we can remove Python 2.6 compatibility in a 7.8.x release so this would need to become 7.9. We still have Cylc in use on RHEL 6 systems so would have to stick to 7.8. Unless there is demand from others, I suggest we make this an 8.0 change.

@dpmatthews, legacy RHEL 6 machines are not the only ones that use python 2.6 for running cylc on our site. Just saying...

hjoliver commented:
> Unless there is demand from others, I suggest we make this an 8.0 change.

Then all users for the next year (ish) will have to live with Jinja2 "incorrect context" bug, which is moderately serious (wrong line numbers are often worse than none, when debugging).

So I think putting to 7.9 would be better, for those not stuck on Python 2.6.  We don't expect many more changes on 7.x so maintaining two branches for subsequent fixes will be easy enough.

dpmatthews commented:
> So I think putting to 7.9 would be better, for those not stuck on Python 2.6. We don't expect many more changes on 7.x so maintaining two branches for subsequent fixes will be easy enough.

Fair enough. I assume this pull request will need to be changed to point to a new 7.9.x branch. The installation guide will also need an update to reflect the Python 2.7 requirement.

kinow commented:
@oliver-sanders 

>Curious, 2.11.1 claims to have 2.7 support. What aspect of the build was breaking? Was is our tests (e.g. pycodestyle) or something functional?

Just this functional test. The output of their traceback changed a bit, so it wasn't matching the output correctly.

kinow commented on CHANGES.md:
```diff
@@ -31,6 +31,10 @@ pages**.
+### Fixes
+
+[#3502](https://github.com/cylc/cylc-flow/pull/3502) - Update to jinja2
+2.11.1, fixing Jinja2 error where validation shows incorrect context.
```
Sure, no problem. Will wait for the new milestone & branch to be created and will rebase and update the commit. Thanks

kinow replied:
Done

oliver-sanders commented:
> Just this functional test. The output of their traceback changed a bit, so it wasn't matching the output correctly.

Sorry the context of that comment got skewed. I was referring to the removal of the `async` files.

kinow commented:
> 
> 
> > Just this functional test. The output of their traceback changed a bit, so it wasn't matching the output correctly.
> 
> Sorry the context of that comment got skewed. I was referring to the removal of the `async` files.

Ah, that one. Travis still has the logs: https://travis-ci.org/kinow/cylc-flow/jobs/645797353

I think it was due to some functional or unit test loading that file?

hjoliver commented:
@kinow I've created the new 7.9.x branch, so you can re-target this PR.

kinow commented:
- Updated PR title
- Rebased against 7.9.x (thanks @hjoliver !) with GitHub UI
- Updated changelog (solving a request from @oliver-sanders to update it, thanks!)
- Push-forced, so now Travis should be busy for a while running functional tests

hjoliver commented:
I'll merge this if it passes the functional tests again (it should) ... we've all agreed in principle, and it doesn't change any files outside of `lib/jinja2/` ...
